Tasks/Gradle/CodeAnalysis/Common/PmdTool.ts

```diff
@@ -19,147 +20,53 @@ import tl = require('vsts-task-lib/task');
-            return null;
-        }
-
-        tl.debug(`[CA] Found ${xmlReports.length} xml reports for module ${output.moduleName}`)
-        return this.buildAnalysisResultFromModule(xmlReports, output.moduleName);
-    }
-
-    private getBuildReportDir(output: ModuleOutput) {
         switch (this.buildOutput.buildEngine) {
             case BuildEngine.Maven:
                 return path.join(output.moduleRoot);
             case BuildEngine.Gradle:
                 return path.join(output.moduleRoot, 'reports', 'pmd');
             default:
-                throw new Error('not supported');
+                throw new Error('No such build engine ' + this.buildOutput.buildEngine);
         }
-    }
 
-    private buildAnalysisResultFromModule(xmlReports: string[], moduleName: string): AnalysisResult {
-
-        let analysisResult: AnalysisResult = null;
-        let fileCount: number = 0;
-        let violationCount: number = 0;
-        let artifacts: string[] = [];
-
-        for (var xmlReport of xmlReports) {
-
-            var pmdXmlFileContents = fs.readFileSync(xmlReport, 'utf-8');
-            var result = this.parseXmlReport(pmdXmlFileContents, xmlReport, moduleName);
-
-            if (result) {
-                violationCount += result[0];
-                fileCount += result[1];
-                artifacts.push(xmlReport);
-                var htmlReport = this.findHtmlReport(xmlReport);
-
-                if (htmlReport) {
-                    artifacts.push(htmlReport);
-                }
-            }
-        }
-
-        return new AnalysisResult(PmdTool.Name, moduleName, artifacts, violationCount, fileCount);
     }
 
-    private findHtmlReport(xmlReport: string): string {
-
-        // expecting to find an html report with the same name
-        var reportName = path.basename(xmlReport, '.xml');
-        var dirName = path.dirname(xmlReport);
-
-        var htmlReports = glob.sync(path.join(dirName, '**', reportName + '.html'));
-
-        if (htmlReports.length > 0) {
-            return htmlReports[0];
-        }
-
-        return null;
-    }
-
-    private parseXmlReport(pmdXmlFileContents: string, xmlReport: string, moduleName: string) {
-
+    /**
+    * Report parser that extracts the number of affected files and the number of violations from a report
+    *
+    * @returns a tuple of [affected_file_count, violation_count]
+    */
+    protected parseXmlReport(xmlReport: string, moduleName: string): [number, number] {
         let fileCount = 0;
         let violationCount = 0;
 
-        xml2js.parseString(pmdXmlFileContents, (err, data) => {
+        var reportContent = fs.readFileSync(xmlReport, 'utf-8');
+        xml2js.parseString(reportContent, (err, data) => {
             // If the file is not XML, or is not from PMD, return immediately
-            if (!data || !data.pmd) { 
+            if (!data || !data.pmd) {
```

Comment:
Good point - !x will check for null, undefined but also false, 0 and "". However given the underlying data types, this seems fine.